flax/linen/recurrent.py

```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
+  time_major: bool = False
+  return_carry: bool = False
+  unroll: int = 1
+  variable_axes: Mapping[lift.CollectionFilter,lift.InOutScanAxis] = FrozenDict()
+  variable_broadcast: lift.CollectionFilter = 'params'
+  variable_carry: lift.CollectionFilter = False
+  split_rngs: Mapping[lift.PRNGSequenceFilter, bool] = FrozenDict({'params': False})
+
+  def __call__(
+    self,
+    inputs: jax.Array,
+    *,
+    initial_carry: Optional[Carry] = None,
+    init_key: Optional[random.KeyArray] = None,
+    segmentation_mask: Optional[Array] = None,
+    return_carry: Optional[bool] = None,
+    time_major: Optional[bool] = None
+  ) -> Union[Output, Tuple[Carry, Output]]:
+    """
+    Applies the RNN to the inputs.
+
+    ``__call__`` allows you to optionally override some attributes like ``return_carry``
+    and ``time_major`` defined in the constructor.
+
+    Arguments:
+      inputs: the input sequence.
+      initial_carry: the initial carry, if not provided it will be initialized
+        using the cell's :meth:`RNNCellBase.initialize_carry` method.
+      init_key: a PRNG key used to initialize the carry, if not provided
+        ``jax.random.PRNGKey(0)`` will be used. Most cells will ignore this
+        argument.
+      segmentation_mask: an integer array of shape ``(*batch, time)`` indicating
+        which elements are part of the sequence and which are padding elements.
+      return_carry: if ``return_carry=False`` (default) only the output sequence is returned,
+        else it will return a tuple of the final carry and the output sequence.
+      time_major: if ``time_major=False`` (default) it will expect inputs with shape
+        ``(*batch, time, *features)``, else it will expect inputs with shape ``(time, *batch, *features)``.
+    Returns:
+      if ``return_carry=False`` (default) only the output sequence is returned,
+      else it will return a tuple of the final carry and the output sequence.
+    """
+
+    if return_carry is None:
+      return_carry = self.return_carry
+
+    if time_major is None:
+      time_major = self.time_major
+
+    # Infer the number of batch dimensions from the input shape.
+    # Cells like ConvLSTM have additional spatial dimensions.
+    num_features_dims = 1 if isinstance(self.cell_size, int) else len(self.cell_size)
+    time_axis = 0 if time_major else inputs.ndim - num_features_dims - 1
+    if time_major:
+      batch_dims = inputs.shape[1:-num_features_dims]
+    else:
+      batch_dims = inputs.shape[:time_axis]
+
+    carry: Carry
+    if initial_carry is None:
+      if init_key is None:
+        init_key = random.PRNGKey(0)
+      carry = self.cell.initialize_carry(
+        init_key, batch_dims=batch_dims, size=self.cell_size)
+    else:
+      carry = initial_carry
+
+    def scan_fn(
+      cell: RNNCellBase, carry: Carry, x: Array
+    ) -> Union[Tuple[Carry, Array], Tuple[Carry, Tuple[Carry, Array]]]:
+      carry, y = cell(carry, x)
+
+      if segmentation_mask is None:
```

Comment:
Could you add a comment why you need this distinction here?

Reply:
Its a small optimization, if we have no segmentation mask we don't need to have the carry as an output. The reason we have the carry as an output is because its faster to select the last carry out side the scan loop than have an `jnp.where` on each step.

I think we should expand this to:

```python
if segmentation_mask is None or not return_carry:
```